### PR TITLE
Advanced Building Blocks - Enumerable Methods

### PR DESCRIPTION
**Pull request by ferbaco86**
On this project, I've applied all the knowledge gained by reading the material on the use of yield, proc, loops, arrays, method definitions by recreating the .each, .each_with_index, .select, .all?, .any?, .none?, .count, .map and .inject Ruby methods.

**Project milestones**
- [x] Recreate some ruby enumerable methods.
- [x] Apply problem-solving skills, dividing the whole problem into smaller parts.

**General requirements:**

- [x] Add a descriptive README file to your repository
- [x] Use meaningful names for your Pull Requests and add a short summary of its content
- [x] Use descriptive commit messages
- [x] Do not commit files that are not related to the project. Instead, add them to .gitignore file.
- [x] Do not commit debugging instructions

**GitFlow:**

- [x] Create a development branch and set it as a default branch in your repository.
- [x] Create a feature-branch branch (give it a descriptive name)
- [x] Use Rubocop to validate your code.

**Project-specific**

- [x] Recreate the .each, .each_with_index, .select, .all?, .any?, .none?, .count, .map and .inject Ruby methods.
- [x] Create .multiply_els method to test .inject
- [x] Extends .map to accept procs

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good Progress :clap: 
You have done a good job with this project on a first attempt. We only have two failing edge cases to fix before we can merge and proceed. Check my comments for suggestions.

### Martin Afani

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    arr
+  end
+
+  def my_each_with_index
+    return enum_for(:my_each_with_index) unless block_given?
+
+    my_each do |item|
+      self.class == Hash ? yield(item, find_index(item)) : yield(item, index(item))
+    end
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] calls the given block once for each element in self
```ruby 
array = Array.new(100){rand(0...9)}
my_each_output = ''
block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
array.each_with_index(&block)
each_output = my_each_output.dup
my_each_output = ''
array.my_each_with_index(&block)
my_each_output == each_output # should return true if my_each_with_index works as expected
```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    if block_given?
+      temp = args[0] if args.length == 1
+      temp = arr.shift unless args.nil? || args.length == 1
+      arr.my_each { |item| temp = yield(temp, item) }
+    else
+      temp = arr.shift unless args.nil? || args.length == 2
+      arr.my_each { |item| temp = temp.send(args[0].to_s, item) }
+    end
+    temp
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) # should return true
```

**Comment by ferbaco86** [comment]
Hi; 
Thanks for the good comment, but I did not find the comments for suggestions you are referring to on your review.

File: enumerables.rb (hunk 1 above)
Comment (ferbaco86) [reply enumerables.rb]: Fixed.
Thanks for the comments!

File: enumerables.rb (hunk 2 above)
Comment (ferbaco86) [reply enumerables.rb]: Thanks for the comment!
The test is returning true

**Comment by Anna-Myzukina** [comment]
# Status: [changes required]


### Good job with the project. Few steps and your project can be approved:
`Implementation details`


* Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history from [Pathwright]()
screenshot 


#### Please check the ruby doc for these [methods](https://ruby-doc.org/core-2.6.5/Enumerable.html) behaviour. Method to look out for is the following:

* my_inject

### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*
*Happy coding * <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
```
Comment (Anna-Myzukina) [inline enumerables.rb]: my_inject identical and returns the same thing as ruby's [inject](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). 


* combines all elements of enum by applying a binary operation, specified by a block
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7

Comment (ferbaco86) [reply enumerables.rb]: Thanks for the commentaries.
The issue has been resoved

**Comment by ferbaco86** [comment]
Implementation details

    Comments: It’s better to delete the commented code. If you need to review a previous version of your code, use git history from Pathwright
> Hi, thanks for the suggestions. The commented lines are only to disable and enable some rubocop rules

**Review by mosesogwo — APPROVED** [review]
**Hi @ferbaco86, you are fit to :rocket:. Your project is approved.**

You can now merge this pull request.

### Status: Approved :heavy_check_mark:

**Pull request merged**